File: 0645-set-mismatch/0645-set-mismatch.java
class Solution {
    public int[] findErrorNums(int[] arr) {
       int i = 0;
       while(i < arr.length){
           int val = arr[i];
           if(val != arr[val - 1]) Swap(arr, i , val - 1);
           else i++;
       } 
       for(int j = 0; j < arr.length; j++){
           if(arr[j] != j + 1) return new int[] {arr[j], j + 1};
       }
       return new int[] {-1, -1};
    }
    public void Swap(int[] arr, int one, int two){
        int temp = arr[one];
        arr[one] = arr[two];
        arr[two] = temp;
    }
}

File: 0645-set-mismatch/0645-set-mismatch.cpp
class Solution {
public:
    vector<int> findErrorNums(vector<int>& nums) {
        vector<int> ans(2, 0);
        map<int, int>m;
        for(int i : nums) m[i]++;
        for(int i = 1; i <= nums.size(); i++){
            if(m[i] == 2) {
                ans[0] = i;
            }
            if(m[i] == 0) ans[1] = i;
        }
        return ans;
    }
};